python/cudf/udf_cpp/groupby/function.cu

```diff
@@ -0,0 +1,329 @@
+    local_min = min(local_min, data[idx]);
+  }
+
+  cuda::atomic_ref<T, cuda::thread_scope_device> ref{block_min};
+  ref.fetch_min(local_min, cuda::std::memory_order_relaxed);
+
+  block.sync();
+
+  return block_min;
+}
+
+template <typename T>
+__device__ int64_t BlockIdxMax(T const* data, int64_t* index, int64_t size)
+{
+  auto block = cooperative_groups::this_thread_block();
+
+  __shared__ T block_max;
+  __shared__ int64_t block_idx_max;
+
+  auto local_max = []() {
+    if constexpr (std::is_floating_point_v<T>) { return -std::numeric_limits<T>::max(); }
+    return std::numeric_limits<T>::min();
+  }();
+  auto local_idx_max = std::numeric_limits<int64_t>::max();
+
+  if (block.thread_rank() == 0) {
+    block_max     = local_max;
+    block_idx_max = local_idx_max;
+  }
+  block.sync();
+
+#pragma unroll
+  for (int64_t idx = block.thread_rank(); idx < size; idx += block.size()) {
+    auto const current_data = data[idx];
+    if (current_data > local_max) {
+      local_max     = current_data;
+      local_idx_max = index[idx];
+    }
+  }
+
+  cuda::atomic_ref<T, cuda::thread_scope_device> ref{block_max};
+  ref.fetch_max(local_max, cuda::std::memory_order_relaxed);
+  block.sync();
+
+  cuda::atomic_ref<int64_t, cuda::thread_scope_device> ref_idx{block_idx_max};
+  if (local_max == block_max) { ref_idx.fetch_min(local_idx_max, cuda::std::memory_order_relaxed); }
+  block.sync();
+
+  return block_idx_max;
+}
+
+template <typename T>
+__device__ int64_t BlockIdxMin(T const* data, int64_t* index, int64_t size)
+{
+  auto block = cooperative_groups::this_thread_block();
+
+  __shared__ T block_min;
+  __shared__ int64_t block_idx_min;
+
+  auto local_min     = std::numeric_limits<T>::max();
+  auto local_idx_min = std::numeric_limits<int64_t>::max();
+
+  if (block.thread_rank() == 0) {
+    block_min     = local_min;
+    block_idx_min = local_idx_min;
+  }
+  block.sync();
+
+#pragma unroll
+  for (int64_t idx = block.thread_rank(); idx < size; idx += block.size()) {
+    auto const current_data = data[idx];
+    if (current_data < local_min) {
+      local_min     = current_data;
+      local_idx_min = index[idx];
+    }
+  }
+
+  cuda::atomic_ref<T, cuda::thread_scope_device> ref{block_min};
+  ref.fetch_min(local_min, cuda::std::memory_order_relaxed);
+  block.sync();
+
+  cuda::atomic_ref<int64_t, cuda::thread_scope_device> ref_idx{block_idx_min};
+  if (local_min == block_min) { ref_idx.fetch_min(local_idx_min, cuda::std::memory_order_relaxed); }
+  block.sync();
+
+  return block_idx_min;
+}
+
+extern "C" {
+#define make_definition(name, cname, type)                                                   \
+  __device__ int name##_##cname(int64_t* numba_return_value, type* const data, int64_t size) \
+  {                                                                                          \
+    *numba_return_value = name<type>(data, size);                                            \
```

Comment:
Should this write to the shared output again be guarded by a condition on thread_idx == 0? (the call to the function can't be, but this is again a (benign?) write race).

Reply:
Yeah, but when I do something like:

```
extern "C" {
#define make_definition(name, cname, type, return_type)                                                   \
  __device__ int name##_##cname(return_type* numba_return_value, type* const data, int64_t size) \
  {                                                                                                \
    return_type temp = name<type>(data, size);                    \
    if (threadIdx.x == 0) *numba_return_value = temp;           \
    __syncthreads();                                                                   \
    return 0;                                                                                 \
  }
```

The result returned is always 0 for some reason. I am not sure why.

Reply:
> The result returned is always 0 for some reason.

To clarify, this only happens to `idxmin/idxmax` functions and others seem to work fine.
```cuda
extern "C" {
#define make_definition_idx(name, cname, type)                                   \
  __device__ int name##_##cname(                                                 \
    int64_t* numba_return_value, type* const data, int64_t* index, int64_t size) \
  {                                                                              \
    auto const res = name<type>(data, index, size);                              \
    if (threadIdx.x == 0) {                                                      \
      *numba_return_value = res;                                                 \
      printf("val: %lld\n", *numba_return_value);                                \
    }                                                                            \
    __syncthreads();                                                             \
    return 0;                                                                    \
  }
```
The CPP output looks correct via `printf` thus I'm not sure how those data are used on the python side.

Reply:
I think that you need to also add a thread guard in `templates.py` otherwise the threads that didn't write to their output value could overwrite thread 0:

```
--- a/python/cudf/cudf/core/udf/templates.py
+++ b/python/cudf/cudf/core/udf/templates.py
@@ -74,5 +74,6 @@ def _kernel(offset, out, index, {input_columns}, {extra_args}):
 
 {group_initializers}
 
-        out[block_id] = f_(dataframe_group, {extra_args})
+        if cuda.threadIdx.x == 0:
+                out[block_id] = f_(dataframe_group, {extra_args})
```

Reply:
> I think that you need to also add a thread guard in templates.py

:+1: It passes all tests but I'm not sure that's what we really want. IIUC, it seems that only thread 0 invokes function `f_` in this case.

Another issue I'm still trying to figure out is that by using `__syncthreads()`, threads of the same block are supposed to return the same value on the cuda side but it's not the case.

Reply:
This _should_ be fixed in https://github.com/rapidsai/cudf/pull/11452/commits/afd094943874713107a47ec786c93e0cfd25a6c2

Reply:
Done